File: packages/java/src/test/java/fyi/sererr/CaptureTest.java
package fyi.sererr;

import org.junit.jupiter.api.Test;

import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

/** Capture semantics: chain ordering, mechanism IDs, frame extraction. */
class CaptureTest {

    /** Single error → 1-entry chain, exception_id 0, parent_id 0. */
    @Test
    void singleErrorOneEntryChain() {
        var e = new RuntimeException("solo");
        List<CapturedError> chain = Capture.capture(e, "RuntimeException", "v1", "host");
        assertThat(chain).hasSize(1);
        var mech = chain.get(0).mechanism().orElseThrow();
        assertThat(mech.exceptionId()).isZero();
        assertThat(mech.parentId()).isZero();
        assertThat(mech.type()).isEqualTo("generic");
        assertThat(mech.handled()).isTrue();
    }

    /** Cause chain → most-causal-first; IDs stamped positionally. */
    @Test
    void threeDeepChain() {
        var inner = new RuntimeException("inner");
        var middle = new RuntimeException("middle", inner);
        var outer = new RuntimeException("outer", middle);
        List<CapturedError> chain = Capture.capture(outer, "RuntimeException", "v1", "host");
        assertThat(chain).hasSize(3);
        assertThat(chain.get(0).message()).isEqualTo("inner");
        assertThat(chain.get(1).message()).isEqualTo("middle");
        assertThat(chain.get(2).message()).isEqualTo("outer");
        assertThat(chain.get(0).mechanism().orElseThrow().exceptionId()).isZero();
        assertThat(chain.get(0).mechanism().orElseThrow().parentId()).isZero();
        assertThat(chain.get(1).mechanism().orElseThrow().exceptionId()).isEqualTo(1);
        assertThat(chain.get(1).mechanism().orElseThrow().parentId()).isZero();
        assertThat(chain.get(2).mechanism().orElseThrow().exceptionId()).isEqualTo(2);
        assertThat(chain.get(2).mechanism().orElseThrow().parentId()).isEqualTo(1);
    }

    /** Originating caught error (outermost) is the LAST entry. */
    @Test
    void outermostIsLast() {
        var inner = new RuntimeException("inner");
        var outer = new RuntimeException("outer", inner);
        List<CapturedError> chain = Capture.capture(outer, "RuntimeException", "v1", "host");
        assertThat(chain.get(chain.size() - 1).message()).isEqualTo("outer");
    }

    /** Frames are non-empty and most-recent-first (throw site is first). */
    @Test
    void framesMostRecentFirst() {
        var e = new RuntimeException("x");
        List<CapturedError> chain = Capture.capture(e, "RuntimeException", "v1", "host");
        var frames = chain.get(0).frames();
        assertThat(frames).isNotEmpty();
        // Top frame should be from this test class (throw site).
        assertThat(frames.get(0).module()).contains("fyi.sererr.CaptureTest");
    }

    /** Suppressed (try-with-resources) exceptions flatten as additional chain entries. */
    @Test
    void suppressedFlattens() {
        var primary = new RuntimeException("primary");
        var suppressed = new RuntimeException("from-close");
        primary.addSuppressed(suppressed);
        List<CapturedError> chain = Capture.capture(primary, "RuntimeException", "v1", "host");
        // Primary + one suppressed = 2 entries.
        assertThat(chain).hasSize(2);
        assertThat(chain).extracting(CapturedError::message)
                .containsExactlyInAnyOrder("primary", "from-close");
    }

    /** Self-referential cause is handled (visited-set breaks the cycle). */
    @Test
    void selfReferentialCauseDoesNotLoop() {
        var e = new RuntimeException("loopy");
        // JVM forbids initCause(self) at construction; we'd reflectively
        // do it but the simpler check: getCause() == this returns null
        // from the JDK. We verify cycles via two-node ring instead.
        var a = new RuntimeException("a");
        var b = new RuntimeException("b", a);
        try {
            a.initCause(b);
        } catch (IllegalStateException ignored) {
            // initCause may fail if cause already set; that's fine for
            // this test — the no-cycle case is sufficient.
            return;
        }
        List<CapturedError> chain = Capture.capture(b, "RuntimeException", "v1", "host");
        // a → b is a cycle; capture must terminate. We accept any
        // finite chain length so long as it doesn't loop.
        assertThat(chain).isNotEmpty();
        assertThat(chain.size()).isLessThanOrEqualTo(8);
    }

    /** captureFrames() works directly on a Throwable without building a chain. */
    @Test
    void captureFramesDirect() {
        var e = new RuntimeException("x");
        List<StackFrame> frames = Capture.captureFrames(e);
        assertThat(frames).isNotEmpty();
    }

    /** Null throwable returns an empty chain (no crash). */
    @Test
    void nullThrowableEmptyChain() {
        assertThat(Capture.capture(null, "T", "v1", "host")).isEmpty();
        assertThat(Capture.captureFrames(null)).isEmpty();
    }

    /** is_app heuristic: stdlib prefixes are non-app, user code is app. */
    @Test
    void isAppFrameHeuristic() {
        assertThat(Capture.isAppFrame("java.util.HashMap")).isFalse();
        assertThat(Capture.isAppFrame("kotlin.collections.MapsKt")).isFalse();
        assertThat(Capture.isAppFrame("com.example.MyService")).isTrue();
        assertThat(Capture.isAppFrame("")).isTrue();
    }
}

File: packages/kotlin/src/test/kotlin/fyi/sererr/kotlin/CaptureTest.kt
package fyi.sererr.kotlin

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

/**
 * Kotlin surface contract for `capture` / `toCapturedErrorChain`.
 *
 * The semantic guarantees are inherited from the Java implementation
 * — these tests pin the Kotlin-idiomatic shape and the type-alias
 * re-exports.
 */
class CaptureTest {

    /** Top-level `capture(throwable, ...)` delegates to the Java impl. */
    @Test
    fun topLevelCapture() {
        val e = RuntimeException("solo")
        val chain = capture(e, "RuntimeException", "v1", "host")
        assertEquals(1, chain.size)
        val mech = chain[0].mechanism().orElseThrow()
        assertEquals(0, mech.exceptionId())
        assertEquals(0, mech.parentId())
    }

    /** Extension `Throwable.toCapturedErrorChain` produces the same shape. */
    @Test
    fun extensionFunction() {
        val inner = RuntimeException("inner")
        val outer = RuntimeException("outer", inner)
        val chain = outer.toCapturedErrorChain("RuntimeException", "v1", "host")
        assertEquals(2, chain.size)
        // most-causal-first: inner is at [0], outer at [1]
        assertEquals("inner", chain[0].message())
        assertEquals("outer", chain[1].message())
    }

    /** Frames from a captured error survive coroutine sanitization. */
    @Test
    fun framesNonEmpty() {
        val e = RuntimeException("with-frames")
        val chain = capture(e, "RuntimeException", "v1", "host")
        assertTrue(chain[0].frames().isNotEmpty(), "expected non-empty frames")
    }

    /** Type aliases point at the Java records (object identity check). */
    @Test
    fun typeAliasesPointAtJavaRecords() {
        val ce: CapturedError = fyi.sererr.CapturedError.empty()
        assertNotNull(ce)
        val sf: StackFrame = fyi.sererr.StackFrame.empty()
        assertNotNull(sf)
        val em: ExceptionMechanism = fyi.sererr.ExceptionMechanism.empty()
        assertNotNull(em)
    }

    /** Coroutine frame recovery is a safe no-op on plain Throwables. */
    @Test
    fun coroutineRecoveryNoOp() {
        val e = RuntimeException("plain")
        val recovered = recoverCoroutineFrames(e)
        // Same throwable returned (recovery sanitizes in place).
        assertTrue(recovered === e, "expected same throwable instance")
    }

    /** Null throwable in Java capture is empty list; Kotlin function is non-null Throwable so we just test via java. */
    @Test
    fun emptyChainOnNullThrowable() {
        // Kotlin's `capture` requires non-null, so we test the Java path
        // directly to confirm parity.
        val chain = fyi.sererr.Capture.capture(null, "T", "v", "h")
        assertEquals(0, chain.size)
    }

    /** DebugInfo extension matches the Java adapter output. */
    @Test
    fun debugInfoExtension() {
        val entry = fyi.sererr.CapturedError.builder()
            .type("E").message("m").build()
        val di = listOf(entry).toDebugInfo()
        assertEquals("E: m", di.detail())
    }

    /** Capture on a throwable with explicit cause chain. */
    @Test
    fun chainedThrowable() {
        val a = RuntimeException("a")
        val b = RuntimeException("b", a)
        val c = RuntimeException("c", b)
        val chain = capture(c, "RuntimeException", "v1", "host")
        assertEquals(3, chain.size)
        assertEquals("a", chain[0].message())
        assertEquals("c", chain[2].message())
    }

    /** Java's IllegalStateException can be initCaused — Kotlin smoke test. */
    @Test
    fun initCauseAllowedAtConstruction() {
        // Kotlin doesn't impose extra restrictions; we just confirm the
        // capture surface accepts JVM-standard chained throwables.
        assertFailsWith<IllegalStateException> {
            error("by-design")
        }.also {
            val chain = it.toCapturedErrorChain("Test", "v1", "host")
            assertEquals(1, chain.size)
            assertEquals("by-design", chain[0].message())
        }
    }
}
